cranelift/peepmatic/src/linearize.rs

```diff
@@ -0,0 +1,831 @@
+                if let Some(path_b) = lhs_id_to_path.get_first_occurrence(id) {
+                    debug_assert!(path != path_b);
+                    (
+                        linear::MatchOp::Eq {
+                            path_a: path,
+                            path_b,
+                        },
+                        Some(1),
+                    )
+                } else {
+                    (linear::MatchOp::Nop, None)
+                }
+            }
+            Pattern::Operation(op) => (linear::MatchOp::Opcode { path }, Some(op.operator as u32)),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use peepmatic_runtime::{
+        integer_interner::IntegerId,
+        linear::{Action::*, MatchOp::*},
+        operator::Operator,
+        r#type::{BitWidth, Kind, Type},
+    };
+
+    macro_rules! linearizes_to {
+        ($name:ident, $source:expr, $make_expected:expr $(,)* ) => {
+            #[test]
+            fn $name() {
+                let buf = wast::parser::ParseBuffer::new($source).expect("should lex OK");
+
+                let opts = match wast::parser::parse::<Optimizations>(&buf) {
+                    Ok(opts) => opts,
+                    Err(mut e) => {
+                        e.set_path(std::path::Path::new(stringify!($name)));
+                        e.set_text($source);
+                        eprintln!("{}", e);
+                        panic!("should parse OK")
+                    }
+                };
+
+                assert_eq!(
+                    opts.optimizations.len(),
+                    1,
+                    "`linearizes_to!` only supports a single optimization; split the big test into \
+                     multiple small tests"
+                );
+
+                if let Err(mut e) = crate::verify(&opts) {
+                    e.set_path(std::path::Path::new(stringify!($name)));
+                    e.set_text($source);
+                    eprintln!("{}", e);
+                    panic!("should verify OK")
+                }
+
+                let mut paths = PathInterner::new();
+                let mut p = |p: &[u8]| paths.intern(Path::new(&p));
+
+                let mut integers = IntegerInterner::new();
+                let mut i = |i: u64| integers.intern(i);
+
+                #[allow(unused_variables)]
+                let expected = $make_expected(&mut p, &mut i);
+                dbg!(&expected);
+
+                let actual = linearize_optimization(&mut paths, &mut integers, &opts.optimizations[0]);
+                dbg!(&actual);
+
+                assert_eq!(expected, actual);
+            }
+        };
+    }
+
+    linearizes_to!(
+        mul_by_pow2_into_shift,
+        "
+(=> (when (imul $x $C)
+          (is-power-of-two $C))
+    (ishl $x $C))
+        ",
+        |p: &mut dyn FnMut(&[u8]) -> PathId, i: &mut dyn FnMut(u64) -> IntegerId| {
+            linear::Optimization {
+                increments: vec![
```

Comment:
Maybe avoid duplication by moving the `linear::Optimization` construction into the macro, leaving this closure to just return a Vec. Also you can try to cast the closure to `fn(&mut dyn FnMut(&[u8]) -> PathId, &mut dyn FnMut(u64) -> IntegerId) -> Vec<linear::Increment>` before calling it. This may help type inference enough to not need explicit type annotations here.

Reply:
Nice idea!